File: frontend/src/utils/nlpEngine.js
function isNumericVal(v) {
  return typeof v === 'number' && !isNaN(v);
}

function isNumericList(values) {
  if (values.length === 0) return false;
  return values.every(v => isNumericVal(v));
}

function isDatetimeStr(val) {
  if (typeof val !== 'string') return false;
  const s = val.trim();
  if (s.length >= 10 && ((s[4] === '-' && s[7] === '-') || (s[4] === '/' && s[7] === '/'))) {
    return true;
  }
  return false;
}

export function generateOfflineSummary(records, anomaliesCount) {
  if (!records || records.length === 0) return "No data processed yet.";
  
  const columns = Object.keys(records[0]);
  let summary = "### Pipeline Insights & Execution Summary (Offline Engine)\n\n";
  summary += `- Dataset Health: Processing completed. Checked ${records.length} rows. Cleaned duplicated rows and resolved missing values.\n`;
  summary += `- Anomaly Count: Found ${anomaliesCount} anomalous events across the dataset.\n\n`;
  
  summary += "#### Key Column Breakdown:\n";
  for (const col of columns) {
    if (['is_anomaly', 'anomaly_reason'].includes(col)) continue;
    
    const vals = records.map(r => r[col]).filter(v => v !== null && v !== undefined);
    const numericVals = vals.map(v => Number(v)).filter(v => !isNaN(v));
    
    // Check if the original column type is numerical (all values are numbers)
    const isNum = vals.length > 0 && vals.every(v => typeof v === 'number');
    
    if (isNum) {
      const meanVal = numericVals.reduce((a, b) => a + b, 0) / numericVals.length;
      const maxVal = Math.max(...numericVals);
      const minVal = Math.min(...numericVals);
      summary += `- ${col}: Numerical. Average is ${meanVal.toFixed(2)} (Range: ${minVal.toFixed(2)} to ${maxVal.toFixed(2)}).\n`;
    } else if (vals.some(v => isDatetimeStr(v))) {
      summary += `- ${col}: Chronological timeline parsed.\n`;
    } else {
      // Categorical
      const nonEmpty = vals.map(v => String(v)).filter(s => s !== "");
      if (nonEmpty.length > 0) {
        const counts = {};
        for (const v of nonEmpty) {
          counts[v] = (counts[v] || 0) + 1;
        }
        const sorted = Object.entries(counts).sort((a, b) => b[1] - a[1]).slice(0, 3);
        const countsStr = sorted.map(([k, v]) => `'${k}': ${v}`).join(", ");
        summary += `- ${col}: Categorical. Primary distributions: ${countsStr}.\n`;
      }
    }
  }
  
  if (anomaliesCount > 0) {
    summary += "\n#### Detected Anomalies:\n";
    let count = 0;
    for (let idx = 0; idx < records.length; idx++) {
      const r = records[idx];
      if (String(r.is_anomaly).toLowerCase() === 'true' || r.is_anomaly === 1 || r.is_anomaly === true) {
        const reason = r.anomaly_reason || 'Threshold exceeded';
        summary += `- Row ${idx}: ${reason}\n`;
        count++;
        if (count >= 5) break;
      }
    }
  }
  
  return summary;
}

export function runOfflineQA(records, question, chatHistory = []) {
  if (!records || records.length === 0) return "No data processed yet.";
  
  let q = question.toLowerCase().trim();
  const columns = Object.keys(records[0]);
  
  const anomIndices = [];
  for (let idx = 0; idx < records.length; idx++) {
    const isAnom = String(records[idx].is_anomaly).toLowerCase() === 'true' || records[idx].is_anomaly === 1 || records[idx].is_anomaly === true;
    if (isAnom) {
      anomIndices.push(idx);
    }
  }
  const anomCount = anomIndices.length;

  // Contextual parsing for conversational flow
  let hasAnomalyContext = false;
  let activeCol = null;

  for (const msg of chatHistory.slice(-3)) {
    const text = String(msg.text).toLowerCase();
    if (['anomaly', 'anomalies', 'outlier', 'outliers', 'flagged', 'flag', 'spike', 'spikes', 'jump', 'jumps'].some(k => text.includes(k))) {
      hasAnomalyContext = true;
    }
    for (const col of columns) {
      if (['is_anomaly', 'anomaly_reason'].includes(col)) continue;
      if (text.includes(col.toLowerCase())) {
        activeCol = col;
      }
    }
  }

  if (hasAnomalyContext && ['why', 'explain', 'reason', 'detail', 'details', 'them', 'show', 'spikes', 'spike'].some(k => q.includes(k))) {
    if (!['anomaly', 'anomalies', 'outlier', 'outliers', 'flagged', 'flag'].some(k => q.includes(k))) {
      q += " anomalies";
    }
  }

  if (activeCol && ['average', 'mean', 'avg', 'maximum', 'max', 'highest', 'peak', 'minimum', 'min', 'lowest', 'sum', 'total'].some(k => q.includes(k))) {
    if (!columns.some(col => col !== 'is_anomaly' && col !== 'anomaly_reason' && q.includes(col.toLowerCase()))) {
      q += ` ${activeCol.toLowerCase()}`;
    }
  }

  // 1. Anomalies queries
  if (['anomaly', 'anomalies', 'outlier', 'outliers', 'flagged', 'flag', 'spike', 'spikes', 'jump', 'jumps'].some(k => q.includes(k))) {
    if (['why', 'reason', 'explain', 'detail', 'details'].some(k => q.includes(k))) {
      if (anomCount === 0) return "No anomalies were detected in the dataset.";
      let response = `Here are the explanations for the detected anomalies (Total: ${anomCount}):\n\n`;
      let shown = 0;
      for (const idx of anomIndices) {
        const reason = records[idx].anomaly_reason || 'Threshold exceeded';
        response += `- Row ${idx}: ${reason}\n`;
        shown++;
        if (shown >= 10) break;
      }
      if (anomCount > 10) {
        response += `\n(Showing first 10 of ${anomCount} anomalies)`;
      }
      return response;
    } else {
      return `There are ${anomCount} anomalies detected in the dataset. You can view them highlighted on the dashboard or query details about them by asking 'why did it spike?' or 'explain anomalies'.`;
    }
  }

  // 2. Aggregations (average, max, min, mean, sum)
  for (const col of columns) {
    if (['is_anomaly', 'anomaly_reason'].includes(col)) continue;
    const colLower = col.toLowerCase();
    if (q.includes(colLower)) {
      const colVals = records.map(r => r[col]).filter(v => v !== null && v !== undefined);
      const isNum = colVals.length > 0 && colVals.every(v => typeof v === 'number');
      
      if (isNum) {
        const floatVals = colVals.map(Number);
        if (['average', 'mean', 'avg'].some(k => q.includes(k))) {
          const val = floatVals.reduce((a, b) => a + b, 0) / floatVals.length;
          return `The average value of ${col} is ${val.toFixed(4)}.`;
        } else if (['maximum', 'max', 'highest', 'peak'].some(k => q.includes(k))) {
          const val = Math.max(...floatVals);
          return `The maximum value of ${col} is ${val.toFixed(4)}.`;
        } else if (['minimum', 'min', 'lowest'].some(k => q.includes(k))) {
          const val = Math.min(...floatVals);
          return `The minimum value of ${col} is ${val.toFixed(4)}.`;
        } else if (['sum', 'total'].some(k => q.includes(k))) {
          const val = floatVals.reduce((a, b) => a + b, 0);
          return `The sum total of ${col} is ${val.toFixed(4)}.`;
        }
      }
    }
  }

  // 3. Categorical distributions
  for (const col of columns) {
    if (['is_anomaly', 'anomaly_reason'].includes(col)) continue;
    const colLower = col.toLowerCase();
    if (q.includes(colLower) || (colLower === 'status' && q.includes('status')) || (colLower === 'category' && q.includes('category'))) {
      const colVals = records.map(r => String(r[col])).filter(s => s !== null && s !== undefined && s !== "");
      const unique = new Set(colVals);
      if (unique.size < 25) {
        const counts = {};
        for (const v of colVals) counts[v] = (counts[v] || 0) + 1;
        const sorted = Object.entries(counts).sort((a, b) => b[1] - a[1]);
        
        let response = `Here is the distribution breakdown for the ${col} column:\n\n`;
        const totalValid = colVals.length;
        for (const [k, v] of sorted) {
          const percentage = totalValid > 0 ? (v / totalValid) * 100 : 0;
          response += `- ${k}: ${v} records (${percentage.toFixed(1)}%)\n`;
        }
        return response;
      }
    }
  }

  // 4. Text query keyword search
  const match = q.match(/["'](.*?)["']/);
  if (match) {
    const searchTerm = match[1].toLowerCase();
    const textCols = columns.filter(col => !['ticket_id', 'client_ip', 'device_id', 'is_anomaly', 'anomaly_reason'].includes(col));
    for (const col of textCols) {
      const matches = [];
      for (let idx = 0; idx < records.length; idx++) {
        const valStr = String(records[idx][col] || '').toLowerCase();
        if (valStr.includes(searchTerm)) {
          matches.push({ idx, text: records[idx][col] });
        }
      }
      if (matches.length > 0) {
        let response = `Found ${matches.length} records where ${col} contains '${searchTerm}':\n\n`;
        for (const m of matches.slice(0, 5)) {
          response += `- Row ${m.idx}: ${m.text}\n`;
        }
        if (matches.length > 5) {
          response += `\n(Showing 5 of ${matches.length} matching records)`;
        }
        return response;
      }
    }
  }

  // 5. Row count
  if (['row', 'record', 'how many', 'count', 'size'].some(k => q.includes(k))) {
    return `The dataset currently has ${records.length} processed rows and ${columns.length} clean data columns.`;
  }

  return (
    "I analyzed your request using the local offline parser, but couldn't resolve a precise database action.\n\n" +
    "Here are examples of questions I can answer directly:\n" +
    "- 'How many anomalies were found?'\n" +
    "- 'Why did temperature spike?' (explanations of anomalies)\n" +
    "- 'What is the average temperature?' (or humidity, vibration, response_time_ms)\n" +
    "- 'What is the highest vibration?'\n" +
    "- 'Show status code distribution'\n" +
    "- 'Show category breakdown'\n\n" +
    "Tip: Configure an LLM API key (Gemini, Hugging Face, or OpenAI) in the dashboard settings to unlock full natural language reasoning and custom reporting!"
  );
}

async function queryLLMEngine(provider, modelName, apiKey, prompt, systemInstruction) {
  if (provider === 'gemini') {
    try {
      const model = modelName || 'gemini-1.5-flash';
      const url = `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${apiKey}`;
      const payload = {
        contents: [{ parts: [{ text: prompt }] }],
        systemInstruction: systemInstruction ? { parts: [{ text: systemInstruction }] } : undefined
      };
      
      const r = await fetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
      });
      
      if (!r.ok) {
        const errJson = await r.json().catch(() => ({}));
        throw new Error(errJson?.error?.message || `HTTP ${r.status}`);
      }
      const data = await r.json();
      return data?.candidates?.[0]?.content?.parts?.[0]?.text || "No text generated.";
    } catch (e) {
      return `Gemini API Error: ${e.message}`;
    }
  } else if (provider === 'openai') {
    try {
      const model = modelName || 'gpt-4o-mini';
      const url = 'https://api.openai.com/v1/chat/completions';
      const messages = [];
      if (systemInstruction) {
        messages.push({ role: 'system', content: systemInstruction });
      }
      messages.push({ role: 'user', content: prompt });
      
      const r = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${apiKey}`
        },
        body: JSON.stringify({ model, messages })
      });
      
      if (!r.ok) {
        const errJson = await r.json().catch(() => ({}));
        throw new Error(errJson?.error?.message || `HTTP ${r.status}`);
      }
      const data = await r.json();
      return data?.choices?.[0]?.message?.content || "No response generated.";
    } catch (e) {
      return `OpenAI API Error: ${e.message}`;
    }
  } else if (provider === 'anthropic') {
    try {
      const model = modelName || 'claude-3-haiku-20240307';
      const url = 'https://api.anthropic.com/v1/messages';
      
      const r = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'x-api-key': apiKey,
          'anthropic-version': '2023-06-01',
          'dangerouslyAllowBrowser': 'true' // In front-end context
        },
        body: JSON.stringify({
          model,
          max_tokens: 1024,
          system: systemInstruction || undefined,
          messages: [{ role: 'user', content: prompt }]
        })
      });
      
      if (!r.ok) {
        const text = await r.text();
        throw new Error(`HTTP ${r.status}: ${text}`);
      }
      const data = await r.json();
      return data?.content?.[0]?.text || "No content generated.";
    } catch (e) {
      return `Claude API Error: ${e.message}`;
    }
  } else if (provider === 'huggingface') {
    try {
      const model = modelName || 'meta-llama/Meta-Llama-3-8B-Instruct';
      const url = `https://api-inference.huggingface.co/models/${model}`;
      const combinedPrompt = `<|system|>\n${systemInstruction || 'You are a data assistant.'}\n<|user|>\n${prompt}\n<|assistant|>\n`;
      
      const r = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${apiKey}`
        },
        body: JSON.stringify({
          inputs: combinedPrompt,
          parameters: { max_new_tokens: 512, return_full_text: false }
        })
      });
      
      if (!r.ok) {
        const text = await r.text();
        throw new Error(`HTTP ${r.status}: ${text}`);
      }
      const data = await r.json();
      if (Array.isArray(data) && data.length > 0) {
        return data[0].generated_text || "";
      }
      return JSON.stringify(data);
    } catch (e) {
      return `Hugging Face API Error: ${e.message}`;
    }
  } else if (provider === 'ollama') {
    try {
      const endpoint = apiKey || 'http://localhost:11434';
      const url = `${endpoint}/api/generate`;
      
      const r = await fetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          model: modelName || 'llama3',
          prompt: systemInstruction ? `${systemInstruction}\n\n${prompt}` : prompt,
          stream: false
        })
      });
      
      if (!r.ok) throw new Error(`HTTP ${r.status}`);
      const data = await r.json();
      return data?.response || "";
    } catch (e) {
      return `Ollama Error: ${e.message}`;
    }
  }
  return "Error: Unknown LLM provider configuration.";
}

function getDatasetContextSummary(records) {
  const totalRows = records.length;
  const anomalies = records.filter(r => String(r.is_anomaly).toLowerCase() === 'true' || r.is_anomaly === 1 || r.is_anomaly === true);
  const anomCount = anomalies.length;
  const columns = totalRows > 0 ? Object.keys(records[0]).join(', ') : '';
  
  let summary = `Dataset statistics:\n`;
  summary += `- Total Rows: ${totalRows}\n`;
  summary += `- Anomalous Rows: ${anomCount}\n`;
  summary += `- Columns: ${columns}\n\n`;
  
  summary += "Sample records:\n";
  summary += JSON.stringify(records.slice(0, 5), null, 2);
  summary += "\n\n";
  
  if (anomCount > 0) {
    summary += "Sample anomaly records:\n";
    summary += JSON.stringify(anomalies.slice(0, 5), null, 2);
    summary += "\n\n";
  }
  
  return summary;
}

export async function getPipelineSummary(records, anomaliesCount, provider, modelName, apiKey) {
  if (!provider || provider === 'offline' || !apiKey) {
    return generateOfflineSummary(records, anomaliesCount);
  }
  
  const context = getDatasetContextSummary(records);
  const systemInst = "You are SmartStream's Executive AI Insights engine. Your goal is to analyze the processed operational dataset context and write a premium, compelling executive summary. Identify patterns, key averages, and highlight specific anomalies.";
  const prompt = `Please analyze the following dataset context and write a comprehensive, professional summary of your findings.\nMake it structured with markdown headings, bullet points, and high-impact conclusions.\n\n${context}`;
  
  return queryLLMEngine(provider, modelName, apiKey, prompt, systemInst);
}

export async function answerDatasetQuestion(records, question, chatHistory, provider, modelName, apiKey) {
  if (!provider || provider === 'offline' || !apiKey) {
    return runOfflineQA(records, question, chatHistory);
  }
  
  const context = getDatasetContextSummary(records);
  const historyStr = chatHistory.slice(-6).map(m => `${m.role.toUpperCase()}: ${m.text}`).join('\n');
  
  const systemInst = "You are SmartStream Q&A, an advanced analytical agent. You answer user questions about the ingested dataset. Use the dataset context provided to answer accurately. If they ask for averages, counts, or anomalies, perform the mental calculation based on the sample records or the statistics provided. Keep responses structured and useful.";
  const prompt = `Here is the dataset context:\n${context}\n\nChat History:\n${historyStr}\nUser Question: ${question}\n\nWrite your response to the user's question now:`;
  
  return queryLLMEngine(provider, modelName, apiKey, prompt, systemInst);
}

File: netlify/functions/nlp_engine.py
import os
import re
import json
import html
import requests
import google.generativeai as genai
from openai import OpenAI
from collections import Counter

def is_numeric_list(values):
    """
    Check if a list of values is numeric.
    """
    if not values:
        return False
    # Exclude booleans since isinstance(True, int) is True in Python
    return all(isinstance(v, (int, float)) and not isinstance(v, bool) for v in values if v is not None)

def is_datetime_str(val):
    """
    Check if a string matches a standard date format.
    """
    if not isinstance(val, str):
        return False
    val = val.strip()
    # Simple check for YYYY-MM-DD or YYYY/MM/DD
    if len(val) >= 10 and ((val[4] == '-' and val[7] == '-') or (val[4] == '/' and val[7] == '/')):
        return True
    return False

def generate_offline_summary(records, anomalies_count):
    """
    Generate a detailed rule-based summary when no LLM key is configured.
    Strips all markdown bolding/asterisks as requested.
    """
    if not records:
        return "No data processed yet."
        
    columns = list(records[0].keys())
    
    summary = "### Pipeline Insights & Execution Summary (Offline Engine)\n\n"
    summary += f"- Dataset Health: Processing completed. Checked {len(records)} rows. Cleaned duplicated rows and resolved missing values.\n"
    summary += f"- Anomaly Count: Found {anomalies_count} anomalous events across the dataset.\n\n"
    
    summary += "#### Key Column Breakdown:\n"
    for col in columns:
        if col in ['is_anomaly', 'anomaly_reason']:
            continue
            
        vals = [r.get(col) for r in records if r.get(col) is not None]
        if is_numeric_list(vals):
            # Calculate metrics
            float_vals = [float(v) for v in vals]
            mean_val = sum(float_vals) / len(float_vals)
            max_val = max(float_vals)
            min_val = min(float_vals)
            summary += f"- {col}: Numerical. Average is {mean_val:.2f} (Range: {min_val:.2f} to {max_val:.2f}).\n"
        elif any(is_datetime_str(v) for v in vals):
            summary += f"- {col}: Chronological timeline parsed.\n"
        else:
            # Categorical distribution
            non_empty = [str(v) for v in vals if v != ""]
            if non_empty:
                counts = Counter(non_empty).most_common(3)
                counts_str = ", ".join([f"'{k}': {v}" for k, v in counts])
                summary += f"- {col}: Categorical. Primary distributions: {counts_str}.\n"
                
    if anomalies_count > 0:
        summary += "\n#### Detected Anomalies:\n"
        count = 0
        for idx, r in enumerate(records):
            if str(r.get('is_anomaly', '')).lower() in ('true', '1', 'yes'):
                reason = r.get('anomaly_reason', 'Threshold exceeded')
                summary += f"- Row {idx}: {reason}\n"
                count += 1
                if count >= 5:
                    break
                    
    return summary

def run_offline_qa(records, question, chat_history=[]):
    """
    Heuristics-based natural language parser to query a pure-Python records list.
    Strips all markdown bolding/asterisks from the output.
    """
    if not records:
        return "No data processed yet."
        
    q = question.lower().strip()
    columns = list(records[0].keys())
    
    # Identify anomaly status in rows
    anom_indices = []
    for idx, r in enumerate(records):
        is_anom = str(r.get('is_anomaly', '')).lower() in ('true', '1', '1.0', 'yes')
        if is_anom:
            anom_indices.append(idx)
            
    anom_count = len(anom_indices)

    # Contextual parsing for conversational flow
    has_anomaly_context = False
    active_col = None

    for msg in chat_history[-3:]:
        text = msg.get('text', '').lower()
        if any(k in text for k in ['anomaly', 'anomalies', 'outlier', 'outliers', 'flagged', 'flag', 'spike', 'spikes', 'jump', 'jumps']):
            has_anomaly_context = True
        for col in columns:
            if col in ['is_anomaly', 'anomaly_reason']:
                continue
            if col.lower() in text:
                active_col = col

    # If recent context was about anomalies, and current question is a follow-up, inject anomaly keyword
    if has_anomaly_context and any(k in q for k in ['why', 'explain', 'reason', 'detail', 'details', 'them', 'show', 'spikes', 'spike']):
        if not any(k in q for k in ['anomaly', 'anomalies', 'outlier', 'outliers', 'flagged', 'flag']):
            q += " anomalies"

    # If recent context was about a specific column, and current question requests an aggregation without column name
    if active_col and any(k in q for k in ['average', 'mean', 'avg', 'maximum', 'max', 'highest', 'peak', 'minimum', 'min', 'lowest', 'sum', 'total']):
        if not any(col.lower() in q for col in columns if col not in ['is_anomaly', 'anomaly_reason']):
            q += f" {active_col.lower()}"

    # 1. Anomaly count & spikes queries
    if any(k in q for k in ['anomaly', 'anomalies', 'outlier', 'outliers', 'flagged', 'flag', 'spike', 'spikes', 'jump', 'jumps']):
        if any(k in q for k in ['why', 'reason', 'explain', 'detail', 'details']):
            if anom_count == 0:
                return "No anomalies were detected in the dataset."
            response = f"Here are the explanations for the detected anomalies (Total: {anom_count}):\n\n"
            shown = 0
            for idx in anom_indices:
                reason = records[idx].get('anomaly_reason', 'Threshold exceeded')
                response += f"- Row {idx}: {reason}\n"
                shown += 1
                if shown >= 10:
                    break
            if anom_count > 10:
                response += f"\n(Showing first 10 of {anom_count} anomalies)"
            return response
        else:
            return f"There are {anom_count} anomalies detected in the dataset. You can view them highlighted on the dashboard or query details about them by asking 'why did it spike?' or 'explain anomalies'."

    # 2. Aggregations (average, max, min, mean, sum)
    for col in columns:
        if col in ['is_anomaly', 'anomaly_reason']:
            continue
        col_lower = col.lower()
        if col_lower in q:
            # Check if column values are numeric
            col_vals = [r.get(col) for r in records if r.get(col) is not None]
            if is_numeric_list(col_vals):
                float_vals = [float(v) for v in col_vals]
                if any(k in q for k in ['average', 'mean', 'avg']):
                    val = sum(float_vals) / len(float_vals)
                    return f"The average value of {col} is {val:.4f}."
                elif any(k in q for k in ['maximum', 'max', 'highest', 'peak']):
                    val = max(float_vals)
                    return f"The maximum value of {col} is {val:.4f}."
                elif any(k in q for k in ['minimum', 'min', 'lowest']):
                    val = min(float_vals)
                    return f"The minimum value of {col} is {val:.4f}."
                elif any(k in q for k in ['sum', 'total']):
                    val = sum(float_vals)
                    return f"The sum total of {col} is {val:.4f}."

    # 3. Categorical breakdown queries
    for col in columns:
        if col in ['is_anomaly', 'anomaly_reason']:
            continue
        col_lower = col.lower()
        if col_lower in q or (col_lower == 'status' and 'status' in q) or (col_lower == 'category' and 'category' in q):
            col_vals = [str(r.get(col)) for r in records if r.get(col) is not None and r.get(col) != ""]
            unique_vals = set(col_vals)
            if len(unique_vals) < 25:
                counts = Counter(col_vals)
                response = f"Here is the distribution breakdown for the {col} column:\n\n"
                total_valid = len(col_vals)
                for k, v in counts.most_common():
                    percentage = (v / total_valid) * 100 if total_valid > 0 else 0
                    response += f"- {k}: {v} records ({percentage:.1f}%)\n"
                return response

    # 4. Text/Keyword searches
    text_cols = [c for c in columns if c not in ['ticket_id', 'client_ip', 'device_id', 'is_anomaly', 'anomaly_reason']]
    for col in text_cols:
        # Check if question asks to search for something in quotes
        match = re.search(r'["\'](.*?)[["\']', question)
        if not match:
            # Fallback to direct check if a keyword matches
            pass
        else:
            search_term = match.group(1).lower()
            matching_rows = []
            for idx, r in enumerate(records):
                val_str = str(r.get(col, '')).lower()
                if search_term in val_str:
                    matching_rows.append((idx, r.get(col)))
            if len(matching_rows) > 0:
                response = f"Found {len(matching_rows)} records where {col} contains '{search_term}':\n\n"
                for idx, text in matching_rows[:5]:
                    response += f"- Row {idx}: {text}\n"
                if len(matching_rows) > 5:
                    response += f"\n(Showing 5 of {len(matching_rows)} matching records)"
                return response

    # 5. Row count/Info queries
    if any(k in q for k in ['row', 'record', 'how many', 'count', 'size']):
        clean_col_count = len(columns)
        return f"The dataset currently has {len(records)} processed rows and {clean_col_count} clean data columns."

    # Default fallback guide
    return (
        "I analyzed your request using the local offline parser, but couldn't resolve a precise database action.\n\n"
        "Here are examples of questions I can answer directly:\n"
        "- 'How many anomalies were found?'\n"
        "- 'Why did temperature spike?' (explanations of anomalies)\n"
        "- 'What is the average temperature?' (or humidity, vibration, response_time_ms)\n"
        "- 'What is the highest vibration?'\n"
        "- 'Show status code distribution'\n"
        "- 'Show category breakdown'\n\n"
        "Tip: Configure an LLM API key (Gemini, Hugging Face, or OpenAI) in the dashboard settings to unlock full natural language reasoning and custom reporting!"
    )

def query_llm_engine(provider, model_name, api_key, prompt, system_instruction=None):
    """
    Dispatch query to external LLM provider.
    """
    if provider == 'gemini':
        try:
            genai.configure(api_key=api_key)
            model = genai.GenerativeModel(
                model_name=model_name or 'gemini-1.5-flash',
                system_instruction=system_instruction
            )
            response = model.generate_content(prompt)
            return response.text
        except Exception as e:
            return f"Gemini API Error: {str(e)}"
            
    elif provider == 'openai':
        try:
            client = OpenAI(api_key=api_key)
            messages = []
            if system_instruction:
                messages.append({"role": "system", "content": system_instruction})
            messages.append({"role": "user", "content": prompt})
            
            completion = client.chat.completions.create(
                model=model_name or "gpt-4o-mini",
                messages=messages
            )
            return completion.choices[0].message.content
        except Exception as e:
            return f"OpenAI API Error: {str(e)}"
            
    elif provider == 'ollama':
        try:
            url = f"{api_key or 'http://localhost:11434'}/api/generate"
            payload = {
                "model": model_name or "llama3",
                "prompt": f"{system_instruction}\n\n{prompt}" if system_instruction else prompt,
                "stream": False
            }
            r = requests.post(url, json=payload, timeout=20)
            if r.status_code == 200:
                return r.json().get('response', '')
            else:
                return f"Ollama API Error: Received status code {r.status_code}"
        except Exception as e:
            return f"Ollama Connection Error: Make sure Ollama server is running locally on port 11434. ({str(e)})"
            
    elif provider == 'huggingface':
        try:
            model = model_name or "meta-llama/Meta-Llama-3-8B-Instruct"
            url = f"https://api-inference.huggingface.co/models/{model}"
            headers = {"Authorization": f"Bearer {api_key}"}
            
            combined_prompt = f"<|system|>\n{system_instruction or 'You are a data assistant.'}\n<|user|>\n{prompt}\n<|assistant|>\n"
            
            payload = {
                "inputs": combined_prompt,
                "parameters": {"max_new_tokens": 512, "return_full_text": False}
            }
            
            r = requests.post(url, headers=headers, json=payload, timeout=20)
            if r.status_code == 200:
                res = r.json()
                if isinstance(res, list) and len(res) > 0:
                    return res[0].get('generated_text', '')
                return str(res)
            elif r.status_code == 503:
                return "Hugging Face Error: Model is currently loading, please try again in a few seconds."
            else:
                return f"Hugging Face Inference Error: Status {r.status_code} - {r.text}"
        except Exception as e:
            return f"Hugging Face API Connection Error: {str(e)}"
            
    elif provider == 'anthropic':
        try:
            url = "https://api.anthropic.com/v1/messages"
            headers = {
                "x-api-key": api_key,
                "anthropic-version": "2023-06-01",
                "content-type": "application/json"
            }
            payload = {
                "model": model_name or "claude-3-haiku-20240307",
                "max_tokens": 1024,
                "messages": [{"role": "user", "content": prompt}]
            }
            if system_instruction:
                payload["system"] = system_instruction
                
            r = requests.post(url, headers=headers, json=payload, timeout=25)
            if r.status_code == 200:
                return r.json().get('content', [{}])[0].get('text', '')
            else:
                return f"Claude API Error: Status {r.status_code} - {r.text}"
        except Exception as e:
            return f"Claude API Connection Error: {str(e)}"
            
    return "Error: Unknown LLM provider configuration."

def get_dataset_context_summary(records):
    """
    Format a condensed summary of the dataset for LLM prompts.
    """
    total_rows = len(records)
    anom_records = [r for r in records if str(r.get('is_anomaly', '')).lower() in ('true', '1', 'yes')]
    anom_count = len(anom_records)
    columns = list(records[0].keys()) if records else []
    
    summary = f"Dataset statistics:\n"
    summary += f"- Total Rows: {total_rows}\n"
    summary += f"- Anomalous Rows: {anom_count}\n"
    summary += f"- Columns: {', '.join(columns)}\n\n"
    
    summary += "Sample records:\n"
    summary += json.dumps(records[:5], indent=2)
    summary += "\n\n"
    
    if anom_count > 0:
        summary += "Sample anomaly records:\n"
        summary += json.dumps(anom_records[:5], indent=2)
        summary += "\n\n"
        
    return summary

def get_pipeline_summary(records, anomalies_count, provider=None, model_name=None, api_key=None):
    """
    Generates the plain-English executive summary for the dashboard.
    """
    if not provider or provider == 'offline' or not api_key:
        return generate_offline_summary(records, anomalies_count)
        
    context = get_dataset_context_summary(records)
    system_inst = "You are SmartStream's Executive AI Insights engine. Your goal is to analyze the processed operational dataset context and write a premium, compelling executive summary. Identify patterns, key averages, and highlight specific anomalies."
    prompt = (
        f"Please analyze the following dataset context and write a comprehensive, professional summary of your findings.\n"
        f"Make it structured with markdown headings, bullet points, and high-impact conclusions.\n\n"
        f"{context}"
    )
    
    return query_llm_engine(provider, model_name, api_key, prompt, system_instruction=system_inst)

def answer_dataset_question(records, question, chat_history=[], provider=None, model_name=None, api_key=None):
    """
    Answers a natural language question about the dataset using LLM or offline heuristics.
    """
    if not provider or provider == 'offline' or not api_key:
        return run_offline_qa(records, question, chat_history)
        
    context = get_dataset_context_summary(records)
    
    # Format chat history
    history_str = ""
    for msg in chat_history[-6:]:
        role = msg.get('role', 'user')
        text = msg.get('text', '')
        history_str += f"{role.upper()}: {text}\n"
        
    system_inst = (
        "You are SmartStream Q&A, an advanced analytical agent. You answer user questions about the ingested dataset. "
        "Use the dataset context provided to answer accurately. If they ask for averages, counts, or anomalies, perform "
        "the mental calculation based on the sample records or the statistics provided. Keep responses structured and useful."
    )
    
    prompt = (
        f"Here is the dataset context:\n{context}\n\n"
        f"Chat History:\n{history_str}\n"
        f"User Question: {question}\n\n"
        f"Write your response to the user's question now:"
    )
    
    return query_llm_engine(provider, model_name, api_key, prompt, system_instruction=system_inst)
